key/key_test.go

```diff
@@ -0,0 +1,409 @@
+		result: false,
+	}, {
+		a:      New(map[string]interface{}{"a": 3}),
+		b:      New(map[string]interface{}{"a": 3}),
+		result: true,
+	}}
+
+	for _, tcase := range tests {
+		if tcase.a.Equal(tcase.b) != tcase.result {
+			t.Errorf("Wrong result for case:\na: %#v\nb: %#v\nresult: %#v",
+				tcase.a,
+				tcase.b,
+				tcase.result)
+		}
+	}
+
+	if New("a").Equal(32) == true {
+		t.Error("Wrong result for different types case")
+	}
+}
+
+func TestIsHashable(t *testing.T) {
+	tests := []struct {
+		k interface{}
+		h bool
+	}{{
+		true,
+		true,
+	}, {
+		false,
+		true,
+	}, {
+		uint8(3),
+		true,
+	}, {
+		uint16(3),
+		true,
+	}, {
+		uint32(3),
+		true,
+	}, {
+		uint64(3),
+		true,
+	}, {
+		int8(3),
+		true,
+	}, {
+		int16(3),
+		true,
+	}, {
+		int32(3),
+		true,
+	}, {
+		int64(3),
+		true,
+	}, {
+		float32(3.2),
+		true,
+	}, {
+		float64(3.3),
+		true,
+	}, {
+		"foobar",
+		true,
+	}, {
+		map[string]interface{}{"foo": "bar"},
+		false,
+	}}
+
+	for _, tcase := range tests {
+		if New(tcase.k).IsHashable() != tcase.h {
+			t.Errorf("Wrong result for case:\nk: %#v",
+				tcase.k)
+
+		}
+	}
+}
+
+func TestGetFromMap(t *testing.T) {
+	tests := []struct {
+		k     Key
+		m     map[Key]interface{}
+		v     interface{}
+		found bool
+	}{{
+		k:     New("a"),
+		m:     map[Key]interface{}{New("a"): "b"},
+		v:     "b",
+		found: true,
+	}, {
+		k:     New(uint32(35)),
+		m:     map[Key]interface{}{New(uint32(35)): "c"},
+		v:     "c",
+		found: true,
+	}, {
+		k:     New(uint32(37)),
+		m:     map[Key]interface{}{New(uint32(36)): "c"},
+		found: false,
+	}, {
+		k:     New(uint32(37)),
+		m:     map[Key]interface{}{},
+		found: false,
+	}, {
+		k: New(map[string]interface{}{"a": "b", "c": uint64(4)}),
+		m: map[Key]interface{}{
+			New(map[string]interface{}{"a": "b", "c": uint64(4)}): "foo",
+		},
+		v:     "foo",
+		found: true,
+	}}
+
+	for _, tcase := range tests {
+		v, ok := tcase.k.GetFromMap(tcase.m)
+		if tcase.found != ok {
+			t.Errorf("Wrong retrieval result for case:\nk: %#v\nm: %#v\nv: %#v",
+				tcase.k,
+				tcase.m,
+				tcase.v)
+		} else if tcase.found && !ok {
+			t.Errorf("Unable to retrieve value for case:\nk: %#v\nm: %#v\nv: %#v",
+				tcase.k,
+				tcase.m,
+				tcase.v)
+		} else if tcase.found && !test.DeepEqual(tcase.v, v) {
+			t.Errorf("Wrong result for case:\nk: %#v\nm: %#v\nv: %#v",
+				tcase.k,
+				tcase.m,
+				tcase.v)
+		}
+	}
+}
+
+func TestDeleteFromMap(t *testing.T) {
+	tests := []struct {
+		k Key
+		m map[Key]interface{}
+		r map[Key]interface{}
+	}{{
+		k: New("a"),
+		m: map[Key]interface{}{New("a"): "b"},
+		r: map[Key]interface{}{},
+	}, {
+		k: New("b"),
+		m: map[Key]interface{}{New("a"): "b"},
+		r: map[Key]interface{}{New("a"): "b"},
+	}, {
+		k: New("a"),
+		m: map[Key]interface{}{},
+		r: map[Key]interface{}{},
+	}, {
+		k: New(uint32(35)),
+		m: map[Key]interface{}{New(uint32(35)): "c"},
+		r: map[Key]interface{}{},
+	}, {
+		k: New(uint32(36)),
```

Comment:
This case and the next case are the same test with slightly different data (both positive-bias "off by one" misses). 

Given there is great coverage in the other rows of the table test, let's save a few score nanoseconds by removing one of these two cases :)


```diff
+		k: New(uint32(36)),
+		m: map[Key]interface{}{New(uint32(35)): "c"},
+		r: map[Key]interface{}{New(uint32(35)): "c"},
+	}, {
+		k: New(uint32(37)),
+		m: map[Key]interface{}{New(uint32(36)): "c"},
+		r: map[Key]interface{}{New(uint32(36)): "c"},
+	}, {
```